accepted/2023/sign-signature-provider-plugins.md

```diff
@@ -0,0 +1,524 @@
+
+This default location could be overridden with an environment variable or CLI option.
+
+The directory structure for the `Plugins` directory will contain one subdirectory for each lower-cased plugin package ID.  Each plugin package ID directory will contain a subdirectory for each lower-cased plugin package version.  Each version subdirectory will contain the extracted contents of the corresponding plugin package.  Example:
+
+```text
+Plugins
+├─microsoft.azure.keyvault.sign
+│ ├─0.9.1-beta.23274.1
+│ │ └─<package contents>
+│ └─0.9.1-beta.23530.1
+│   └─<package contents>
+└─<another plugin package ID>
+  └─<package version>
+    └─<package contents>
+```
+
+To identify installed plugins, Sign CLI will simply look in this directory for packages and use the latest version (release or prerelease) using SemVer 2.0.0 versioning rules.
+
+Sign CLI will not maintain any installation state.  How a plugin was installed --- using NuGet client tools, Sign CLI, manual package extraction, or some other method --- is immaterial.  The presence of an extracted package in the directory means it has been installed.
+
+#### Plugin instantiation
+
+A plugin's command in Sign CLI will:
+
+1. Read a plugin's `plugin.json` file.
+1. Find the plugin's entry point that best matches the host (Sign CLI) not the runtime and deviate only if it's a sound fallback.  Examples:
+    1. If a plugin contains net6.0, net7.0, and net8.0 assemblies and Sign CLI's net7.0 assemblies are loaded by the .NET 8 runtime, Sign CLI should load the plugin's net7.0 assemblies.
+    1. If a plugin only had net6.0 and net8.0 assemblies, Sign CLI should load the net6.0 assemblies or report that plugin isn't available for the current runtime.
+1. Load the assembly at the entry point's `filePath` location.
+1. Create an instance of the type `implementationTypeName`.
+1. Cast the instance to an interface defined in the interfaces package with type name `interfaceTypeName`.
+
+As part of this process, Sign CLI will use [`System.Runtime.Loader.AssemblyLoadContext`](https://learn.microsoft.com/dotnet/api/system.runtime.loader.assemblyloadcontext) and [`System.Runtime.Loader.AssemblyDependencyResolver`](https://learn.microsoft.com/dotnet/api/system.runtime.loader.assemblydependencyresolver) to load a plugin assembly and its dependencies strictly from the directory cone of the plugin's entry point assembly.  For example, if the plugin's entry point is `lib/net6.0/Microsoft.Azure.KeyVault.Sign.dll`, then Sign CLI will attempt to resolve assemblies under `lib/net6.0`.
+
+Sign CLI should log:
+
+* high-level plugin loading information at the [`Information`](https://learn.microsoft.com/dotnet/api/microsoft.extensions.logging.loglevel?view=dotnet-plat-ext-7.0#fields) log level.
+* assembly loading details at `Debug` / `Trace` log levels
+* errors at the `Error` log level
+
+### Sign CLI commands
```

Comment:
Should "list plugins" and "remove plugin" commands also exist?

Reply:
The intention was this spec would represent v1 of the feature, which would not have those commands.  Those commands could be added later.

Alternatively, we could spec them now and prioritize implementation after everything else.  (The same could be said about other potential features, but we were trying to expedite v1 by not speccing too much.)